crypto/init.c

```diff
@@ -729,3 +633,46 @@ int OPENSSL_atexit(void (*handler)(void))
     return 1;
 }
 
+/*
+ * Library  destructor (a.k.a. _fini()). Code here is derived
+ * from code found in providers/fips/self_test.c
+ *
+ * We do kind of best effort cleanup. If given platform/compiler
+ * does not support module destructors then its up to application
+ * to explicitly call OPENSSL_cleanup();
+ */
+#define DEF_CALL_CLEANUP_FUN(_func_name)\
+    _func_name(void)			\
+    {					\
+        if (ossl_no_cleanup == 0)	\
+            OPENSSL_cleanup();		\
+    }
+
+#if defined(__GNUC__) && !defined(_AIX)
+
+void ossl_call_cleanup(void);
+void __attribute__((destructor))
+DEF_CALL_CLEANUP_FUN(ossl_call_cleanup)
+
+#elif (defined(__sun) || defined(_AIX)) && !defined(__GNUC__)
+
+void ossl_call_cleanup(void);
+# pragma fini(ossl_call_cleanup)
+void DEF_CALL_CLEANUP_FUN(ossl_call_cleanup)
+
+#elif defined(__hpux)
+
+void ossl_call_cleanup(void);
+# pragma fini "ossl_call_cleanup"
+void DEF_CALL_CLEANUP_FUN(ossl_call_cleanup)
+
+#elif defined(__TANDEM)
+
+/* Method automatically called by the NonStop OS prior to unloading the DLL */
+void DEF_CALL_CLEANUP_FUN(__TERM__cleanup)
+
+#elif !defined(WIN32)
+
+void DEF_CALL_CLEANUP_FUN(_fini)
+
+#endif
```

Comment:
Should this be `_WIN32` instead of `WIN32`? Also it is not clear to me what is happening on Windows - there does not seem to be any implicit cleanup there anymore.

Reply:
crypto/dllmain.c calls `OPENSSL_cleanup()` on its own
```
 41         break;
 42     case DLL_PROCESS_DETACH:
 43         if (ossl_no_cleanup == 0)
 44             OPENSSL_cleanup();
 45         break;
```